### PR TITLE
DROOLS-2877: Update the top level data type, considering changes in an external/nested sub data type

### PR DESCRIPTION
**Pull request by karreiro**
See:

* [DROOLS-2877](https://issues.jboss.org/browse/DROOLS-2877): Update the top level data type, considering changes in an external/nested sub data type

* [DROOLS-2881](https://issues.jboss.org/browse/DROOLS-2881): [DMN Designer] Data type dialog - "Destroy" action in the list view

* [DROOLS-2945](https://issues.jboss.org/browse/DROOLS-2945): DMN Designer - opening data type dialog shows warning in browser console

* [DROOLS-2950](https://issues.jboss.org/browse/DROOLS-2950): [DMN Editor] Custom data-types - ItemDefinitionRecordEngine - The QName instance should lookup the real prefix from the URI

drools-2877

drools-2881

### CODE REVIEW

**Comment by manstis** [comment]
Heads up... I ran (quickly) on top of DROOLS-2941 and got a NPE when dragging a Decision from the palette. Will try again tomorrow (when I'm less rushed).. but if you're still around you might want to check too.

**Comment by karreiro** [comment]
@manstis Thank you for the heads up! :-)

..but, I could not reproduce the issue here, see:
2018-09-18 4 52 58 pm

**Comment by manstis** [comment]
Thanks for investigating @karreiro 

Don't try to hard to replicate. It was the first thing I encountered before anything else.

I'll try better tomorrow!

**Comment by karreiro** [comment]
@manstis ok! Have a nice evening :-)

**Comment by manstis** [comment]
Hi @karreiro here's observations (not reviewed the code yet):

Edit a type (click "Edit" button)
- Change data-type to something different (e.g. was "String" change to "Number")
- Click "X" to cancel change
- Text in dialog shows cancelled change
- Click "Edit" again and the data-type drop-down correctly shows the original data-type

Updated data-type of element
- Changed "tGender" from "string" to "time"
- Saved
- Examined XML
-- Actual: <semantic:typeRef>{http://www.omg.org/spec/FEEL/20140401}time</semantic:typeRef>
-- Expected: <semantic:typeRef>feel:time</semantic:typeRef>

---------------------------------------------------------------
(May be out of scope of changes made)

Select kebab menu next to a nested/child element
- Select "Remove"
- NPE

Edit "tProfiles" (that is of type "tProfile") and shows the child elements (defined in "tProfile")
- Change "Age" from "Number" to something else
- Save
- "tProfiles" looses its child content (i.e although showing as a "Structure" it no longer lists the content of "tProfile")
- "tProfile" Age field remains "Number"

**Comment by manstis** [comment]
File I used is attached.

[DMCommunity Challenge - March 2017.zip](https://github.com/kiegroup/kie-wb-common/files/2396549/DMCommunity.Challenge.-.March.2017.zip)

**Comment by karreiro** [comment]
@mantis Thank you for the feedback (and thank you for the `.dmn` file, it was really helpful).

Ok, about the issues..

> Edit a type (click "Edit" button)

Fixed, see:
cancel-bug


> Updated data-type of element

Fixed, see:
xml


> Select kebab menu next to a nested/child element (of top-level element); Remove it; `Cannot read property 'isTopLevel_1_g$' of undefined` (...)

Fixed, see:
remove-tprofile


> Edit "tProfiles" (that is of type "tProfile") and shows the child elements (defined in "tProfile"); Save it (...)

Fixed, see:
update-tprofiles


> Edit "tProfiles" (that is of type "tProfile") and shows the child elements (defined in "tProfile"); Remove it (...)

Fixed, see:

remove-tprofiles

Thanks again for the review.

**Comment by karreiro** [comment]
@jomarko Thank you for the review.

> Could you please add javadoc at least for new classes and public methods inside? Otherwise it is really hard to do some valuable review.
I added the Javadoc in the main classes, I hope it helps.

> The PR is really big
Yes.

> We use a lot of similar terms: handler/manager, nested/sub
I added some comments to help, but something that can help too is checking where the classes are being used, the package of the class, the pattern that each class implement or even the implementation approach.

For example, the class `DataTypeManager` looks like a _Builder_, however, it's not called a builder because besides encapsulating the object creation, we're encapsulating its interaction with Data Types when they are already created (see the `withRefreshedSubDataTypes(..)` method).

All `*Handlers` (inside of the package `org.kie.workbench.common.dmn.client.editors.types.persistence.handlers`) handle with persistence tasks. Externally, they are called only by the `ItemDefinitionRecordEngine`, which is the class responsible for recording an ItemDefinition considering the changes suffered by a Data Type).

Regarding the concepts "nested" and "sub", I see them as different things. When we're working closer to the View, the nested concept makes more sense, however, when we think how the Data Types related to each other, we can see the `subDataType` expression in the code.

> We are not consistent
> Somewhere we compare data types as a.equals(b), on other place Objects.equals(a, b)
> Somewhere we iterate lists via streams, somewhere via for cycles
I think that we could always use `Objects.equals(a, b)` instead of `a.equals(b)`. If I miss it in some place, we can fix it.
Regarding how we're handling lists in general, `stream`s and `for`s provide similar features, however sometimes one is more readable than another, there's no reason for creating strict rules about this :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/DataTypeHandler.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+import java.util.function.Function;
+
+import org.kie.workbench.common.dmn.client.editors.types.common.DataType;
+import org.kie.workbench.common.dmn.client.editors.types.common.DataTypeManager;
+import org.kie.workbench.common.dmn.client.editors.types.persistence.DataTypeStore;
+import org.kie.workbench.common.dmn.client.editors.types.persistence.ItemDefinitionRecordEngine;
+
+public class DataTypeHandler {
+
+    final DataTypeStore dataTypeStore;
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/DataTypeHandler.java]: This property is being used by the classes that extend `DataTypeHandler` :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/ItemDefinitionUpdateHandler.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        itemDefinition.setName(makeName(dataType));
+    }
+
+    Name makeName(final DataType dataType) {
+        return new Name(dataType.getName());
+    }
+
+    QName makeQName(final DataType dataType) {
+        if (isDefault(dataType.getType())) {
+            return new QName(FEEL.getUri(), dataType.getType());
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/ItemDefinitionUpdateHandler.java]: I think this is fixed now :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeUtils.java [hunk 3]
```diff
@@ -16,72 +16,45 @@
...
-import org.kie.workbench.common.dmn.client.graph.DMNGraphUtils;
+import org.kie.workbench.common.dmn.client.editors.types.persistence.DataTypeStore;
+
+import static java.util.stream.Collectors.toList;
 
 @Dependent
 public class DataTypeUtils {
 
-    private final DataTypeFactory dataTypeFactory;
+    private final ItemDefinitionUtils itemDefinitionUtils;
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeUtils.java]: Fixed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListView.java [hunk 4]
```diff
@@ -96,17 +101,57 @@ public void addSubItems(final DataType dataType,
...
             parent = itemElement;
         }
+
+        showArrowIconIfDataTypeHasChildren(dataType);
+    }
+
+    void hideItemElementIfParentIsCollapsed(final HTMLElement itemElement,
+                                            final Element parent) {
+
+        final boolean isCollapsed = isCollapsed(parent.querySelector("[data-field=\"arrow-button\"]"));
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListView.java]: Fixed :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListView.java [hunk 5]
```diff
@@ -96,17 +101,57 @@ public void addSubItems(final DataType dataType,
...
+    }
+
+    @Override
+    public void removeItem(final DataType dataType) {
+
+        cleanSubTypes(dataType.getUUID());
+
+        final Optional<Element> dataTypeRow = Optional.ofNullable(getDataTypeRow(dataType));
+
+        dataTypeRow.ifPresent(Element::remove);
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListView.java]: Fixed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeSelectView.java [hunk 6]
```diff
@@ -151,17 +155,24 @@ public void setDataType(final DataType dataType) {
         typeText.textContent = "(" + dataType.getType() + ")";
     }
 
-    public void onSelectChange() {
-        presenter.refreshView(typeSelect.value);
+    public void onSelectChange(final JQuerySelectPickerEvent event) {
+
+        final String newValue = event.target.value;
+
+        if (!newValue.equals(getValue())) {
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeSelectView.java]: Fixed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/DataTypeStore.java [hunk 7]
```diff
@@ -41,7 +44,19 @@ public void clear() {
...
+    public int size() {
         return dataTypes.size();
     }
+
+    public List<DataType> getTopLevelDataTypes() {
+        return all().stream().filter(DataType::isTopLevel).collect(Collectors.toList());
+    }
+
+    public List<DataType> all() {
+        return new ArrayList<>(dataTypes.values());
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/DataTypeStore.java]: Fixed :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/ItemDefinitionStore.java [hunk 8]
```diff
@@ -22,7 +22,13 @@
...
 import org.kie.workbench.common.dmn.api.definition.v1_1.ItemDefinition;
+import org.kie.workbench.common.dmn.client.editors.types.DataTypeModal;
 
+/**
+ * Stores all Item Definitions loaded in the {@link DataTypeModal}.
+ * <p>
+ * All entries are indexed by the UUID from the correspondent Data Type.
+ */
 @ApplicationScoped
 public class ItemDefinitionStore {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/ItemDefinitionStore.java]: I was comparing this class with `DataTypeStore` and noticed here we don't have `unIndex` metod. Isn't it issue as `DataTypeStore.unIndex` is used in `DataTypeDestroyHandler#destroy` and `ItemDefinitionDestroyHandler#destroy` doesn't use any `unIndex` method.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/RecordEngine.java [hunk 9]
```diff
@@ -16,11 +16,30 @@
...
+     * @param record
+     * @return returning an array of all affected records by the update operation.
+     */
+    List<T> update(final T record);
 
-    void destroy(T record);
+    /**
+     * Destroys a record
+     * @param record
+     * @return returning an array of all affected records by the update operation.
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/RecordEngine.java]: by the destroy operation

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/DataTypeHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+        doReturn(dataType4).when(handler).parent(dataType5);
+        doReturn(dataType5).when(handler).parent(dataType1);
+        doReturn(dataType6).when(handler).parent(dataType2);
+        doReturn(dataType6).when(handler).parent(dataType3);
+        when(dataTypeStore.getTopLevelDataTypes()).thenReturn(asList(dataType4, dataType6));
+
+        final Optional<DataType> closestTopLevelDataType = handler.getClosestTopLevelDataType(dataType1);
+        final DataType actualDataType = closestTopLevelDataType.orElseThrow(UnsupportedOperationException::new);
+
+        assertEquals(dataType4, actualDataType);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/DataTypeHandlerTest.java]: Would you mind to check also if closest top level data type for dataType3 is dataType6?

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/ItemDefinitionDestroyHandler.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        final Optional<ItemDefinition> itemDefinitionParent = Optional.ofNullable(itemDefinitionStore.get(dataType.getParentUUID()));
+
+        if (itemDefinitionParent.isPresent()) {
+            final ItemDefinition parent = itemDefinitionParent.get();
+
+            if (parent.getTypeRef() == null) {
+                return Optional.of(parent);
+            } else {
+                for (final ItemDefinition itemDefinition : itemDefinitions()) {
+                    if (itemDefinition.getName().getValue().equals(parent.getTypeRef().getLocalPart())) {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/ItemDefinitionDestroyHandler.java]: Please use `Objects.equals`

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/ItemDefinitionUtils.java [hunk 12]
```diff
@@ -50,4 +52,21 @@ public ItemDefinitionUtils(final DMNGraphUtils dmnGraphUtils) {
...
+        final String localPart = typeRef.getLocalPart();
+        final String typeRefPrefix = typeRef.getPrefix();
+        final Optional<String> nsPrefix = getPrefixForNamespaceURI(namespace);
+
+        return nsPrefix
+                .map(prefix -> new QName("", localPart, prefix))
+                .orElseGet(() -> new QName(namespace, localPart, typeRefPrefix));
+    }
+
+    Optional<String> getPrefixForNamespaceURI(final String namespace) {
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/ItemDefinitionUtils.java]: Technically this isn't 100% correct; but I think based on what we expose in our editor it is good enough.

```ItemDefinition``` extends ```DMNModelInstrumentedBase``` and hence can have its own Name Spaces defined (e.g. FEEL URI ```http://www.omg.org/spec/FEEL/20140401``` _could_ map to ```cheese``` in the ```ItemDefinition``` and hence we'd need the ```QName``` to be ```cheese:string```). *However* we do not support editing the Name Spaces for an ```ItemDefinition``` and hence always using ```Definitions``` is good enough :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeManager.java [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+import static java.util.stream.Collectors.toList;
+import static org.kie.soup.commons.validation.PortablePreconditions.checkNotNull;
+import static org.kie.workbench.common.dmn.client.editors.types.common.DataType.TOP_LEVEL_PARENT_UUID;
+import static org.kie.workbench.common.dmn.client.resources.i18n.DMNEditorConstants.DataTypeManager_None;
+import static org.kie.workbench.common.dmn.client.resources.i18n.DMNEditorConstants.DataTypeManager_Structure;
+
+/**
+ * Manages all operations that change a Data Type.
+ * <p>
+ * A Data Type is a plan entity that must simple as simple as possible.
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeManager.java]: Fixed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeManager.java [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+        final Optional<QName> typeRef = Optional.ofNullable(itemDefinition.getTypeRef());
+        return typeRef.isPresent() ? typeRef.get().getLocalPart() : structure();
+    }
+
+    private ItemDefinition getItemDefinition(final DataType dataType) {
+        final Optional<ItemDefinition> itemDefinition = Optional.ofNullable(itemDefinitionStore.get(dataType.getUUID()));
+        return itemDefinition.orElseThrow(() -> new UnsupportedOperationException("The data type must have an indexed ItemDefinition."));
+    }
+
+    private List<DataType> itemDefinitionSubDataTypes() {
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeManager.java]: This method creates sub data types from the `this.itemDefinition` attribute.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeManager.java [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+                    .map(this::createSubDataType)
+                    .collect(toList());
+        } else {
+            return new ArrayList<>();
+        }
+    }
+
+    ItemDefinition getItemDefinitionWithItemComponent(final ItemDefinition itemDefinition) {
+        if (itemDefinition.getTypeRef() != null) {
+            final Optional<ItemDefinition> definition = findByName(itemDefinition.getTypeRef().getLocalPart());
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeManager.java]: Fixed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/common/ItemDefinitionUtilsTest.java [hunk 16]
```diff
@@ -65,6 +70,82 @@ public void testFindByName() {
...
+    }
+
+    @Test
+    public void testNormaliseTypeRefWhenPrefixForNamespaceURIIsPresent() {
+
+        final String expectedLocalPart = "string";
+        final String expectedPrefix = "feel";
+        final String expectedNamespace = "";
+        final String namespaceURI = FEEL.getUri();
+        final QName qName = new QName(namespaceURI, expectedLocalPart);
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/common/ItemDefinitionUtilsTest.java]: @jomarko here we're instantiating the parameter, the _real `QName`_ is instantiated inside of the `normaliseTypeRef` method, and that's we're asserting here :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/common/ElementHelper.java [hunk 17]
```diff
@@ -24,4 +24,8 @@ public static void insertAfter(final Node newNode,
                                    final Node referenceNode) {
         referenceNode.parentNode.insertBefore(newNode, referenceNode.nextSibling);
     }
+
+    public static void remove(final Node element) {
+        element.parentNode.removeChild(element);
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/common/ElementHelper.java]: Thanks :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/ItemDefinitionDestroyHandler.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+                    }
+                }
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    List<ItemDefinition> itemDefinitions() {
+        return dmnGraphUtils.getDefinitions().getItemDefinition();
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/ItemDefinitionDestroyHandler.java]: Fixed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java [hunk 19]
```diff
@@ -232,28 +240,110 @@ public void testCloseEditMode() {
...
+        doReturn(removedDataTypes).when(listItem).removeTopLevelDataTypes(destroyedDataTypes);
+        doReturn(dataType).when(listItem).getDataType();
+
+        listItem.remove();
+
+        verify(dataTypeList).refreshItemsByUpdatedDataTypes(asList(dataType0, dataType3));
+    }
+
+    @Test
+    public void testRemoveTopLevelDataTypes() {
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java]: Fixed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/ItemDefinitionStore.java (hunk 8 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/ItemDefinitionStore.java]: Fixed (The `unIndex` method were created and called in the pertinent places).

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/RecordEngine.java (hunk 9 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/RecordEngine.java]: Fixed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/ItemDefinitionDestroyHandler.java (hunk 11 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/ItemDefinitionDestroyHandler.java]: Fixed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/DataTypeHandlerTest.java (hunk 10 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/DataTypeHandlerTest.java]: Fixed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/ItemDefinitionUtils.java (hunk 12 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/ItemDefinitionUtils.java]: Ok.. thank you, @manstis :-)

**Comment by jomarko** [comment]
Could someone please add description into [DROOLS-2950](https://issues.jboss.org/browse/DROOLS-2950), otherwise I am not able to verify that this PR fixes that jira. Other manual checks are done from my site and found issues are reported.

**Pull request merged**